internal/controller/namespacelabel_controller_test.go

```diff
@@ -0,0 +1,188 @@
+package controller
+
+import (
+	"context"
+	"log"
+	"os"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+
+	labelsv1 "github.com/matanamar10/namespacelabel-operator/api/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+)
+
+var (
+	ctx       context.Context
+	k8sClient client.Client
+	scheme    *runtime.Scheme
+)
+
+var _ = BeforeSuite(func() {
+	By("Initializing test environment")
+	scheme = runtime.NewScheme()
+	Expect(labelsv1.AddToScheme(scheme)).To(Succeed())
+	Expect(corev1.AddToScheme(scheme)).To(Succeed())
+	k8sClient = fake.NewClientBuilder().WithScheme(scheme).Build()
+	ctx = context.Background()
+	if err := os.Setenv("PROTECTED_LABELS", `{"protected-key": "value1", "another-protected-key": "value2"}`); err != nil {
+		log.Fatalf("Failed to set PROTECTED_LABELS environment variable: %v", err)
+	}
+})
+
+func createNamespace(name string) {
+	namespace := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{Name: name},
+	}
+	Expect(k8sClient.Create(ctx, namespace)).To(Succeed())
+}
+
+func deleteNamespace(name string) {
+	namespace := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{Name: name},
+	}
+	Expect(k8sClient.Delete(ctx, namespace)).To(Succeed())
+}
+
+func deleteAllNamespaceLabels() {
+	namespaceLabelList := &labelsv1.NamespacelabelList{}
+	Expect(k8sClient.List(ctx, namespaceLabelList)).To(Succeed())
+	for _, nl := range namespaceLabelList.Items {
+		Expect(k8sClient.Delete(ctx, &nl)).To(Succeed())
+	}
+}
+
+func getNextEvent(recorder *record.FakeRecorder) string {
+	select {
+	case event := <-recorder.Events:
+		return event
+	default:
+		return ""
+	}
+}
+
+var _ = Describe("NamespaceLabel Controller", func() {
+	BeforeEach(func() {
+		By("Creating default namespace")
+		createNamespace("default")
+		deleteAllNamespaceLabels()
+	})
+
+	AfterEach(func() {
+		By("Cleaning up test resources")
+		deleteAllNamespaceLabels()
+		deleteNamespace("default")
+	})
+
+	Context("When reconciling multiple NamespaceLabel resources in the same namespace", func() {
+		const namespaceName = "default"
+
+		It("should apply only non-overlapping labels from multiple NamespaceLabel resources", func() {
```

Comment:
Where are the test cases for `protectedLabels`?

Reply:
i wiil add